src/connectionbenchmark/benchmarktasktransfer.cpp

```diff
@@ -0,0 +1,244 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "benchmarktasktransfer.h"
+#include "constants.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "networkrequest.h"
+#include "uploaddatagenerator.h"
+
+#include <QByteArray>
+#include <QDnsLookup>
+#include <QHostAddress>
+#include <QScopeGuard>
+
+#if !defined(MVPN_DUMMY)
+constexpr const char* MULLVAD_DEFAULT_DNS = "10.64.0.1";
+#endif
+
+namespace {
+Logger logger(LOG_MAIN, "BenchmarkTaskTransfer");
+}
+
+BenchmarkTaskTransfer::BenchmarkTaskTransfer(BenchmarkType type,
+                                             const QUrl& url)
+    : BenchmarkTask("BenchmarkTaskTransfer",
+          Constants::BENCHMARK_MAX_DURATION_TRANSFER),
+      m_type(type),
+      m_dnsLookup(QDnsLookup::A, url.host()),
+      m_url(url) {
+  MVPN_COUNT_CTOR(BenchmarkTaskTransfer);
+
+  connect(this, &BenchmarkTask::stateChanged, this,
+          &BenchmarkTaskTransfer::handleState);
+  connect(&m_dnsLookup, &QDnsLookup::finished, this,
+          &BenchmarkTaskTransfer::dnsLookupFinished);
+}
+
+BenchmarkTaskTransfer::~BenchmarkTaskTransfer() {
+  MVPN_COUNT_DTOR(BenchmarkTaskTransfer);
+}
+
+void BenchmarkTaskTransfer::handleState(BenchmarkTask::State state) {
+  logger.debug() << "Handle state" << state;
+
+  if (state == BenchmarkTask::StateActive) {
+// Start DNS resolution
+#if !defined(MVPN_DUMMY)
+    m_dnsLookup.setNameserver(QHostAddress(MULLVAD_DEFAULT_DNS));
+#endif
+
+#if QT_VERSION >= 0x060400
+#     error Check if QT added support for QDnsLookup::lookup() on Android
+#endif
+
+#if defined(MVPN_ANDROID) || defined(MVPN_WASM)
+    createNetworkRequest();
+#else
+    m_dnsLookup.lookup();
```

Comment:
this is confusing. too many #if/#endif. I guess what we want to do is:

```
#if defined(MVPN_DUMMY) || defined(MVPN_ANDROID) || defined(MVPN_WASM)
 createNetworkRequest();
#else
m_dnsLookup.setNameserver(QHostAddress(MULLVAD_DEFAULT_DNS));
 m_dnsLookup.lookup();
#endif

#if QT_VERSION >= 0x060400
#     error Check if QT added support for QDnsLookup::lookup() on Android
#endif
```

Reply:
Yes — that looks much better.